arch/arm64: Implement ASID support in ARM64 MMU

hx-bora opened the pull request:
Improves a lot context-switch performance.

TLB Invalidating and the nG bit are used conservatively. This could be improved in future work.

Note: MMU configurations changes are known to be error prone. Please have someone review this patch very carefully.

npitre commented:
Thanks for this work.

Some comments:

- Could you provide numbers in the commit log illustrating the context 
  switch improvements? For example, the output from tests/benchmarks/sched 
  before and after this change.

- Please define `next_asid` as a static variable.

- If you remove `is_ptable_active()` you should probably replace those
  `invalidate_tlb_all()` invocations that were conditional on 
  `is_ptable_active()` by an ASID-based TLB invalidate version, otherwise
  performance will likely become worse not better.

- This code:

```
        if ((next_asid >= (1U << VM_ASID_BITS)) || (next_asid == 0)) {
                next_asid = 1;
        }
```

  Testing against 0 is redundant as it will never happen.

povergoing commented on arch/arm64/core/mmu.c:
```diff
@@ -1155,9 +1179,17 @@ int arch_mem_domain_thread_remove(struct k_thread *thread)
 static void z_arm64_swap_ptables(struct k_thread *incoming)
 {
 	struct arm_mmu_ptables *ptables = incoming->arch.ptables;
+	uint64_t curr_ttbr0 = read_sysreg(ttbr0_el1);
```
please use `read_ttbr0_el1()`

povergoing commented:
I am also curious about the improvements :)

npitre commented:
Excellent.

If you could add your test code in tests/kernel/sched/ and then
include those results in the commit log for the ASID support 
that'd be perfect.

Commit:
tests/benchmarks: Add userspace scheduling benchmark

Zephyr already has a scheduling benchmark in tests/benchmarks/sched,
but it only uses kernel threads.
We add an userspace version of it, to exercise memory domains.

Signed-off-by: Henri Xavier <datacomos@huawei.com>

Commit:
arch/arm64: Implement ASID support in ARM64 MMU

Improves context-switch performance.

TLB invalidation and the nG bit are used conservatively. This could
be improved in future work.

Tested with tests/benchmarks/sched_userspace:

BEFORE:
```
Swapping  2 threads: 161562583 cyc & 1000000 rounds ->   1615 ns per ctx
Swapping  8 threads: 161569289 cyc & 1000000 rounds ->   1615 ns per ctx
Swapping 16 threads: 161649163 cyc & 1000000 rounds ->   1616 ns per ctx
Swapping 32 threads: 163487880 cyc & 1000000 rounds ->   1634 ns per ctx
```

AFTER:
```
Swapping  2 threads: 18129207 cyc & 1000000 rounds ->    181 ns per ctx
Swapping  8 threads: 49702891 cyc & 1000000 rounds ->    497 ns per ctx
Swapping 16 threads: 55898650 cyc & 1000000 rounds ->    558 ns per ctx
Swapping 32 threads: 58059704 cyc & 1000000 rounds ->    580 ns per ctx
```

Signed-off-by: Henri Xavier <datacomos@huawei.com>
